webapp/shop/api/ua_contracts/advantage_mapper.py

```diff
@@ -186,3 +209,123 @@ def get_user_subscriptions(self, email: str) -> List[UserSubscription]:
             )
 
         return build_user_subscriptions(user_summary, listings)
+
+    def get_or_create_user_account(
+        self, marketplace, customer_info, captcha_value
+    ) -> Account:
+        try:
+            return self.get_purchase_account(marketplace)
+        except (UnauthorizedError, UAContractsUserHasNoAccount):
+            email = customer_info.get("email", "") if customer_info else ""
+            name = customer_info.get("name", "") if customer_info else ""
+
+            account = self.ensure_purchase_account(
+                marketplace=marketplace,
+                email=email,
+                account_name=name,
+                captcha_value=captcha_value,
+            )
+
+            self.ua_contracts_api.set_authentication_token(account.token)
+            self.ua_contracts_api.set_token_type("Bearer")
```

Comment:
I know this is nasty, but I was unsure how to change the state of the API object for the remaining calls to the API (to use the token received when we created the user's purchase account).

Basically as a guest we use `Beader <token>` authentication and for logged in users its `Macaroon <long-cookie-thing>`.